### DISCUSSION TITLE
How does set prevalue using MQTT work

### DISCUSSION THREAD

**Discussion by danieldownload** [jomjol/AI-on-the-edge-device#2612] [General]
Hi there,

coming from version 15.1.1, being now on 15.3.0, I read about the possibility to set the prevalue via mqtt in case of an read error.

How does this work? I'm using ioBroker and I see the datapoint "set_prevalue", what is needed to do in order to activate the function.

From the feature request page I read the a value of -1 can be set, does this mean that the prevalue is the automatically set to the latest RAW value, in case an error occurs?

Thanks for this great piece of software.

Daniel

**Comment by caco3** [comment]
Did you see the documentation?

https://jomjol.github.io/AI-on-the-edge-device-docs/MQTT-API/#ctrlset_prevalue

**Comment by danieldownload** [comment]
I did but it doesn't seem to work, I still get a lot of errors.

What's wrong with my config?

<img width="1441" alt="Bildschirmfoto 2023-09-19 um 18 27 00" src="https://github.com/jomjol/AI-on-the-edge-device/assets/26095164/c0e9d1ff-9cc7-4a0b-884c-5aa02102dd17">

**Reply by caco3** [reply]
what errors?

**Comment by danieldownload** [comment]
The error, that the pre value is higher than the current value.

Neg. Rate - Read: - Raw: 00067.4449 - Pre: 67.4460

I thought it could be corrected automatically by mqtt.

**Reply by caco3** [reply]
does prevalue have the raw value afterwards?
did you wait a round? The error will only go away then!

**Reply by danieldownload** [reply]
The prevalue didn't change, I waited last time two hours, then I changed the value manually to fix it.

I have this situation every day, gonna make some screenshots next time.

**Comment by danieldownload** [comment]
It just happend again.

<img width="1005" alt="Bildschirmfoto 2023-09-20 um 10 15 30" src="https://github.com/jomjol/AI-on-the-edge-device/assets/26095164/e6af03df-84b8-4914-9c2b-bcd0b14135aa">

<img width="1439" alt="Bildschirmfoto 2023-09-20 um 10 17 28" src="https://github.com/jomjol/AI-on-the-edge-device/assets/26095164/686aaa30-c1c9-47bb-9b56-e731b991dd7c">

**Comment by caco3** [comment]
Not sure why it does not work for you, I just tested and successfully was able to set it to a given value or the latest raw value:
![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/1783586/b3ad94af-8a62-4647-adca-64612fd5e977)

Log:
```
[0d19h04m57s] 2023-09-20T22:27:04	<DBG>	[MQTT IF] handler_set_prevalue called: numbersname: main, value: 606.000000
[0d19h04m57s] 2023-09-20T22:27:05	<INF>	[POSTPROC] SetPreValue: PreValue for main set to 606.000000
```

---

![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/1783586/20cae855-6456-40f7-8eec-9a1ee7441c0a)

Log:
```
[0d19h07m13s] 2023-09-20T22:29:20	<DBG>	[MQTT IF] handler_set_prevalue called: numbersname: main, value: -1.000000
[0d19h07m13s] 2023-09-20T22:29:21	<INF>	[POSTPROC] SetPreValue: PreValue for main set to 604.512320
```
---
In each case, after reloading the Web UI, it showed the right value!

**Reply by danieldownload** [reply]
Interesting behavior, when I change the value to any number in an error state, then this number gets immediately set to the esp32 and shows up, when I then set the value back to -1, it's updating the esp32 also to the last pre-value, which fixes the error.
